### ISSUE TITLE
Cross-library API types

### ISSUE THREAD

**Issue by anderslanglands** [vfx-rs/cppmm#63]
We discussed this briefly on the last call. The problem is how do we want to handle things like:
```
namespace Imf {
class DeepImage {
    DeepImage(Imath::Box2i& dataWindow, /* ... */);
};
}
```
The problem here is that `Imath::Box2i` doesn't exist in OpenEXR and our bindings currently rename the type using the user-specified alias to `Imath_Box2i_t`.

One option would be to somehow supply the binding files for Imath to OpenEXR so that the OpenEXR binding knows about the types and how we've renamed them.

Another, (imo) better, option would be to mangle the C++ type names to create the C struct name, then use typedefs to apply the  "nice" names.

Obviously this means that we need a way of generating a C-compatible mangled name from the C++ name. The simple way would be to just replace all unsupported characters with underscores (which I think we were doing in an earlier version of asttoc anyway).
```
typedef struct {/*...*/} Imath_2_5__Box_Imath_2_5__Vec2_int__t;
typedef Imath_2_5__Box_Imath_2_5__Vec2_int__t Imath_Box2i_t;
```
Obviously that name is ugly, but not quite as hideously ugly as the other option which would be using Clang's mangler to do it which would be something like `_ZN9Imath_2_5E3BoxN9Imath_2_5I4Vec2I3int`.

I think using Clang's mangling would probably be safer and less brittle, but yeah it makes your eyes bleed.

**Comment by luke-titley** [comment]
One thing to add is that asttoc already has the C struct declaration + the typedef. The struct is needed for forward declarations.
```
typedef struct Imath_Box2i_s {
  // data
} Imath_Box2i;
```

I already have a todo to suffix the nice name with _t, but could also change the struct name too.
```
typedef struct Imath_2_5__Box_Imath_2_5__Vec2_int__t {
  // data
} Imath_Box2i_t;
```

Only issue with that is that using the struct everywhere in the headers might make things even nastier.

```
void foo(struct  Imath_2_5__Box_Imath_2_5__Vec2_int__t * input);
```

In which case it should probably look like this.
```
typedef struct Imath_2_5__Box_Imath_2_5__Vec2_int__s {
  // data
} Imath_2_5__Box_Imath_2_5__Vec2_int__t;
typedef Imath_2_5__Box_Imath_2_5__Vec2_int__t Imath_Box2i_t;
```

**Comment by anderslanglands** [comment]
Yep that looks like the way to go.

**Comment by anderslanglands** [comment]
The fix in PR #63 creates some issues with the symbol uniquifying for types like half. Need to take another pass at this.

**Comment by anderslanglands** [comment]
I'm going to close this for now. Just adding binding files for dependencies' types that are in the library's API actually doesn't appear to be an issue. Will open a new issue if we find we need to.
